--- library_utils.c.orig	2012-01-23 10:12:16 +0100
+++ library_utils.c	2012-01-23 10:13:20 +0100
@@ -54,7 +54,7 @@
 #include <io.h>
 #endif
 
-#if !defined (_MSC_VER)
+#if !defined (_MSC_VER) && !defined (_WIN32)
 #include <pwd.h>
 #endif
 
@@ -460,7 +460,7 @@
  * a terminal (and hence won't work from the GUI).
  * @return pointer to character string containing login name.
  */
-#if ! defined (_MSC_VER)
+#if ! defined (_MSC_VER) && ! defined (_WIN32)
 char *ccp4_utils_username(void)
 { 
   static char userid_unknown[] = "unknown";
@@ -653,7 +653,7 @@
  * @param tarray Array containing User and System times.
  * @return Sum of User and System times.
  */
-#if ! defined (_MSC_VER) 
+#if ! defined (_MSC_VER) && ! defined (_WIN32)
 float ccp4_utils_etime (float tarray[2])
 {
   static long clk_tck = 0;
