--- src/Makefile.am.orig	2019-12-02 13:46:07.939246400 +0100
+++ src/Makefile.am	2019-12-02 13:46:18.387844000 +0100
@@ -118,7 +118,7 @@
 stamp-goocanvasenumtypes.h: @REBUILD@ $(goo_canvas_public_headers) Makefile
 	(cd $(srcdir) && $(GLIB_MKENUMS) \
 		--fhead "#ifndef __GOO_CANVAS_ENUM_TYPES_H__\n#define __GOO_CANVAS_ENUM_TYPES_H__\n\n#include <glib-object.h>\n\nG_BEGIN_DECLS\n" \
-		--fprod "/* Enumerations from \"@filename@\" */\n" \
+		--fprod "//* Enumerations from \"@filename@\" */\n" \
 		--vhead "GType @enum_name@_get_type (void);\n#define GOO_TYPE_@ENUMSHORT@ (@enum_name@_get_type())\n\n\n" \
 		--ftail "G_END_DECLS\n\n#endif /* __GOO_CANVAS_ENUM_TYPES_H__ */" \
 		$(libgoocanvas_public_headers)) >> xgen-gtbh \
